src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs

```diff
@@ -293,17 +293,37 @@ static bool CompareTypeLayout(MetadataType type1, MetadataType type2)
                         (layoutMetadata1.Size != layoutMetadata2.Size))
                         return false;
 
-                    if ((explicitLayout) && !(layoutMetadata1.Offsets == null && layoutMetadata2.Offsets == null))
+                    if (explicitLayout)
```

Comment:
We walk all fields on both types starting at line 209 in this file already - I would just add the check there so that we don't need to do this ugly compare-two-IEnumerables dance twice.